unmarshal.go

```diff
@@ -0,0 +1,298 @@
+package apiclient
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"reflect"
+	"strings"
+)
+
+// Unmarshal unmarshals a byte array into a predefined proto.Message.
+func Unmarshal(b []byte, pm proto.Message) error {
+	jsonMap := make(map[string]interface{})
+	if err := json.Unmarshal(b, &jsonMap); err != nil {
+		return err
+	}
+	grrValue, hasValue := jsonMap["value"]
+	grrType, hasType := jsonMap["type"].(string)
+	if hasValue && hasType {
+		return unmarshalTypeValue(pm, grrType, grrValue)
+	} else {
+		return unmarshalMap(pm, jsonMap)
+	}
+	return nil
```

Comment:
I can't provide a full list of types that go through the typeValue- and through the map-unmarshaler, it depends on how the GRR-Response is constructed.

In some cases, e.g. `/api/clients/C.1000000000000000/flows/F:ABCDEF12`, GRR returns the TypeValue construct in the top level:
`{ "age": 0, "type": "ApiFlow", "value": {...}}`

In other cases, e.g. `/api/clients/C.1000000000000000/last-ip`, GRR returns the object directly:
`{ "info": "Internal IP address.", "ip": "192.168.100.42", "status": "INTERNAL" }`